File: backend/app/schemas/corridors.py
from __future__ import annotations

from typing import Literal

from pydantic import AliasChoices, BaseModel, Field


class CorridorAnalysisRequest(BaseModel):
    road_id: str = Field(
        min_length=1,
        max_length=64,
        # Accept camelCase from the frontend (roadId) and snake_case from
        # API docs / curl examples (road_id) without a separate transformer.
        validation_alias=AliasChoices("roadId", "road_id"),
    )
    # Buffer distance in metres around the road geometry used to count nearby
    # features.  Defaults to 30 m (roughly one lane width on each side).
    buffer_meters: int = Field(default=30, ge=0, le=500)


class CorridorAnalysisResponse(BaseModel):
    """
    Corridor planning summary returned by POST /api/corridors/analyze.

    Fields use camelCase to match the frontend TypeScript type (CorridorSummary)
    without requiring an alias on the response side.
    """
    corridorId: str
    roadId: str
    name: str
    knownCurbRamps: int
    possibleMissingCurbCuts: int
    hydrantsNearby: int
    bikeLanesNearby: int
    userAnnotationsNearby: int
    busStopsNearby: int
    parkingConflicts: int
    bikeLaneFeasibility: Literal["Low", "Medium", "High"]
    planningNotes: list[str]

File: frontend/src/types/corridors.ts
/** Corridor planning summary returned by POST /api/corridors/analyze. */
export interface CorridorSummary {
  corridorId: string;
  roadId: string;
  name: string;
  knownCurbRamps: number;
  possibleMissingCurbCuts: number;
  hydrantsNearby: number;
  bikeLanesNearby: number;
  userAnnotationsNearby: number;
  busStopsNearby: number; // Always 0 until CURBO has a documented transit-stop source.
  parkingConflicts: number;
  bikeLaneGaps: number;
  intersectionSafetyConcerns: number;
  annotationsNeedingReview: number;
  bikeLaneFeasibility: "Low" | "Medium" | "High";
  reviewPriority: "Low" | "Medium" | "High";
  reviewSignals: string[];
  dataLimitation: string;
  planningNotes: string[];
}
